File: pipelines/metatranscriptomics/scripts/main_get_sequence_length.pl
#!/usr/local/bin/perl -w
use strict;
use Bio::SeqIO;

my ($read_files, $read_type) = @ARGV;

my $Datapath = "~/CourseData/metagenomics/metatranscriptomics/";
my $Workpath = "";

my $IDfile;
my $infile;
my $infile2;
my $IDfile2;
if ($read_type eq 'rRNA') {
    $IDfile = $Workpath.$read_files."1_qual_all_unique_IDs.txt";
    $infile = $Workpath.$read_files."1_qual_all_unique.fastq";
    $IDfile2 = $Workpath.$read_files."2_qual_all_unique_IDs.txt";
    $infile2 = $Workpath.$read_files."2_qual_all_unique.fastq";
} elsif ($read_type eq 'singletons') {
    $infile = $Workpath.$read_files."1_singletons.fastq";
    $infile2 = $Workpath.$read_files."2_singletons.fastq";
} elsif ($read_type eq 'micro_cds_sub') {
    $infile = $Workpath."microbial_cds_sub.fasta";
} elsif ($read_type eq 'nr_sub') {
    $infile = $Workpath."nr_all_sub.fasta";
}

my ($root, $suff) = $infile =~ m/(.*)\.(.*)/;

my $outfile = $root."_IDs_length.txt";
print "Outputfile: $outfile\n\n";

my %reads;
if (defined $IDfile) {
    open(INPUT, $IDfile) or die "Error opening $IDfile : $!\n";
    while( my $line = <INPUT> ) {
        chomp($line);
        my @values = split(/\t/, $line);
        $reads{$values[0]} = $values[1];
    }
    close INPUT;
    print "number of unique reads:  ".keys( %reads ).".\n\n\n";
}

unlink($outfile);
open(OUTPUT, '>>', $outfile) or die "Error opening $outfile : $!\n";
my $seqIO = Bio::SeqIO->new( -file => $infile, -format => $suff );
while (my $seqobj = $seqIO->next_seq) {
    my $headf = $seqobj->description;
    my $head = $seqobj->display_id;
    my $seq_length = $seqobj->length;
    if ($read_type eq 'rRNA') {
        print OUTPUT "$head\t$reads{$head}\t$seq_length\n";
    } elsif ($read_type eq 'micro_cds_sub') {
        print OUTPUT "$head\t$seq_length\n";
    } elsif ($read_type eq 'nr_sub') {
        print OUTPUT "$head$headf\t$head\t$seq_length\n";
    } else {
        print OUTPUT "$head\t1\t$seq_length\n";
    }
}
close OUTPUT;

if (defined $infile2) {
    my ($root2, $suff2) = $infile2 =~ m/(.*)\.(.*)/;
    my $outfile2 = $root2."_IDs_length.txt";
    print "Outputfile: $outfile2\n\n";

    my %reads;
    if (defined $IDfile) {
        open(INPUT2, $IDfile2) or die "Error opening $IDfile2 : $!\n";
        while( my $line = <INPUT2> ) {
            chomp($line);
            my @values = split(/\t/, $line);
            $reads{$values[0]} = $values[1];
        }
        close INPUT2;
        print "number of unique reads:  ".keys( %reads ).".\n\n\n";
    }

    unlink($outfile2);
    open(OUTPUT2, '>>', $outfile2) or die "Error opening $outfile2 : $!\n";
    my $seqIO = Bio::SeqIO->new( -file => $infile2, -format => $suff2 );
    while (my $seqobj = $seqIO->next_seq) {
        my $head = $seqobj->display_id;
        my $seq_length = $seqobj->length;
        if ($read_type eq 'rRNA') {
            print OUTPUT2 "$head\t$reads{$head}\t$seq_length\n";
        } else {
            print OUTPUT2 "$head\t1\t$seq_length\n";
        }
    }
    close OUTPUT2;
}

exit 0;

File: pipelines/metatranscriptomics/scripts/main_get_sequence_length.py
#!/usr/bin/env python
"""
Read a fastq or fasta and file and write out an ID file containing:
<ID><TAB>[2nd field of id-file<TAB>]<sequence length>

Additional information may be specified by an id-file, which fills in the second field above
"""
import argparse
import os
import json
from Bio import SeqIO


def parse_args():
    arg_parser = argparse.ArgumentParser()
    arg_parser.add_argument('--fastq')
    arg_parser.add_argument('--fasta')
    arg_parser.add_argument('--id-file')
    arg_parser.add_argument('--output')
    return arg_parser.parse_args()


def parse_id_file(id_file, reads):
    """
    Write the "value" field of read based on id_file
    :param reads: list of reads from SeqIO
    :return:
    """
    def get_id(line):
        """
        Get the ID from a line
        """
        return line.strip().split('\t')[0]

    def get_value(line):
        """
        Get the 2nd value from a line
        """
        return line.strip().split('\t')[1]

    id_to_value = {get_id(line): get_value(line) for line in open(id_file)}

    # Set the "value" field of each read
    for read in reads:
        read.value = id_to_value[read.id]

    return reads


def write_output(reads, output_file):
  #  if args.id_file:
  #      def write_read(read):
  #           return '{id}\t{value}\t{len}\n'.format(id=read.id, value=read.value, len=len(read.seq))
  #  else:
  #      def write_read(read):
  #           return '{id}\t{len}\n'.format(id=read.id, len=len(read.seq))

    def write_read(read):
        with_id = '{id}\t{value}\t{len}\n'.format(id=read.id, value=read.value, len=len(read.seq))
        without_id = '{id}\t{len}\n'.format(id=read.id, len=len(read.seq))

        return with_id if args.id_file else without_id

    if('nr_all_sub' in output_file):
        with open(output_file, 'w+') as out:
            # Write header line
            out.write("@id\tdescription\tlength\n")
            for read in reads:
                out.write("{_id}\t{description}\t{length}\n".format(
                                                                    _id=read.id,
                                                                    description=read.description,
                                                                    length=len(read.seq)))
    elif('microbial_cds' in output_file):
        with open(output_file, 'w+') as out:
            # Write header line
            out.write("@id\tlength\n")
            for read in reads:
                out.write("{_id}\t{length}\n".format(_id=read.id,
                                                        length=len(read.seq)))

    else:
        with open(output_file, 'w+') as out:
            # Write header line
            out.write("@id\tlength\n")
            for read in reads:
                out.write("{_id}\t{length}\n".format(_id=read.id,
                                                        length=len(read.seq)))


args = parse_args()
input, format = (args.fastq, 'fastq') if args.fastq else (args.fasta, 'fasta')

reads = list(SeqIO.parse(input, format))

if args.id_file:
    reads = parse_id_file(args.id_file, reads)

write_output(reads, args.output)
